core/src/main/java/com/sequenceiq/cloudbreak/service/stackpatch/ExistingStackPatchService.java

```diff
@@ -0,0 +1,63 @@
+package com.sequenceiq.cloudbreak.service.stackpatch;
+
+import static com.sequenceiq.cloudbreak.util.Benchmark.checkedMeasure;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.StackPatch;
+import com.sequenceiq.cloudbreak.domain.stack.StackPatchType;
+import com.sequenceiq.cloudbreak.repository.StackPatchRepository;
+import com.sequenceiq.flow.core.FlowLogService;
+
+public abstract class ExistingStackPatchService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExistingStackPatchService.class);
+
+    @Inject
+    private StackPatchRepository stackPatchRepository;
+
+    @Inject
+    private FlowLogService flowLogService;
+
+    public boolean isStackAlreadyFixed(Stack stack) {
+        return stackPatchRepository.findByStackAndType(stack, getStackFixType()).isPresent();
+    }
+
+    public void apply(Stack stack) throws ExistingStackPatchApplyException {
+        if (!flowLogService.isOtherFlowRunning(stack.getId())) {
+            try {
+                LOGGER.info("Starting stack {} patching for {}", stack.getResourceCrn(), getStackFixType());
+                checkedMeasure(() -> doApply(stack), LOGGER, "Existing stack patching {} took {} ms for stack: {}.", getStackFixType(), stack.getResourceCrn());
+                LOGGER.info("Stack {} was patched successfully for {}", stack.getResourceCrn(), getStackFixType());
+                stackPatchRepository.save(new StackPatch(stack, getStackFixType()));
+            } catch (ExistingStackPatchApplyException e) {
+                throw e;
+            } catch (Exception e) {
+                String message = String.format("Something unexpected went wrong with stack %s while applying patch %s",
+                        stack.getResourceCrn(), getStackFixType());
+                throw new ExistingStackPatchApplyException(message, e);
+            }
+        } else {
+            throw new ExistingStackPatchApplyException("Another flow is running for stack " + stack.getResourceCrn());
```

Comment:
is the exception really necessary? I mean this is more like business as usual, and we should just log that it didn't happened because a flow was running

Reply:
If we do not throw an exception, the job will be unscheduled. Another option would be to use a boolean to indicate if the apply was successful, would it be a better solution?

Reply:
so in my opinion a better way would be if this `apply` method would return with a result object, which could have error(s) and success, and depending on that the caller side in `ExistingStackPatcherJob#applyStackPatches` could decide if it was successful. At least that how we collect successful and failed operations on FMS side. Exception is an expensive operation in Java world, and I think we already build too many logic on that, like salt polling. So if it's possible I would avoid throwing exceptions if it's not really an error. And a running flow in case of a cluster is just fine. (sure, you would throw an exception if this would be a request on the API, to notify the caller that the call failed.)
but if this would be a too big refactor we can leave this way for now and do it later

Reply:
I see, thank you! I will include this optimization in the follow up ticket we discussed with @pdarvasi